Brushes Support

jsuarezruiz opened the pull request:
### Description of Change ###

Include Brushes support in .NET MAUI.

<img width="1729" alt="netmaui-brushes" src="https://user-images.githubusercontent.com/6755973/110808656-81359b00-8284-11eb-9fe3-356433c604c8.png">

### Platforms Affected ### 

- Core
- iOS
- Android

### PR Checklist ###

- [x] Targets the correct branch 
- [ ] Tests are passing (or failures are unrelated)

jsuarezruiz committed:
Implement Brushes in .NET MAUI

jsuarezruiz committed:
Fix build errors

jsuarezruiz committed:
Merge branch 'main' into brushes

jsuarezruiz committed:
Fixed broken unit tests

jsuarezruiz committed:
Merge branch 'main' into brushes

jsuarezruiz committed:
Merge branch 'main' into brushes

jsuarezruiz committed:
Merge branch 'main' into brushes

jsuarezruiz committed:
Added nullability fixes

jsuarezruiz committed:
Merge branch 'main' into brushes

jsuarezruiz committed:
Fix code format error

jsuarezruiz committed:
Merge branch 'main' into brushes

jsuarezruiz committed:
Merge branch 'main' into brushes

jsuarezruiz committed:
Updated iOS BrushExtensions

jsuarezruiz committed:
Merge branch 'main' into brushes

jsuarezruiz committed:
Fixed build errors

mattleibow committed:
Merge remote-tracking branch 'origin/main' into brushes

mattleibow committed:
try this

mattleibow committed:
try this

mattleibow committed:
that

mattleibow committed:
oops

mattleibow committed:
sample

mattleibow committed:
fix that

mattleibow committed:
Revert some things quick

mattleibow committed:
Merge remote-tracking branch 'origin/main' into brushes

mattleibow committed:
and this

mattleibow committed:
do this for iOS

mattleibow committed:
do android too

mattleibow commented on src/Core/src/Graphics/MauiDrawable.Android.cs:
```diff
@@ -0,0 +1,196 @@
+
+namespace Microsoft.Maui.Graphics
+{
+	public class MauiDrawable : PaintDrawable
```
This drawable only has our 3 types of brushes because we only have 3. This could be split into separate classes and each brush return it. But this is probably not worth it now as it works and is similar to what we have already.

mattleibow commented on src/Core/src/Handlers/Label/LabelHandler.cs:
```diff
@@ -4,6 +4,9 @@ public partial class LabelHandler
 		public static PropertyMapper<ILabel, LabelHandler> LabelMapper = new PropertyMapper<ILabel, LabelHandler>(ViewHandler.ViewMapper)
 		{
+#if __IOS__ || IOS
+			[nameof(ICheckBox.Background)] = MapBackground,
```
Labels on iOS are special in that they can't actually have a layer set. In Forms this "worked" because there was specific logic to check for labels and then apply the brush to the renderer view and not the label inside. Here we use custom logic on `MauiLabel`.

mattleibow commented on src/Core/src/Handlers/View/AbstractViewHandler.MaciOS.cs:
```diff
@@ -13,7 +14,12 @@ public partial class AbstractViewHandler<TVirtualView, TNativeView> : INativeVie
+			{
 				View.Frame = rect.ToCGRect();
+
+				if (VirtualView?.Background?.IsNullOrEmpty() == false)
+					View.UpdateBackgroundLayerFrame();
```
iOS is special in that the layers need to be resize when the control is resized.

mattleibow committed:
Not used

mattleibow commented:
<img width="996" alt="Screenshot 2021-03-27 at 00 17 06" src="https://user-images.githubusercontent.com/1096616/112698173-c61b1d80-8e91-11eb-9fae-849695c4d0af.png">

mattleibow committed:
fix tests

mattleibow committed:
and windows

mattleibow committed:
Unify WinUI versions

mattleibow committed:
Merge branch 'dev/reunion-versions-props' into brushes

mattleibow committed:
final thing

mattleibow committed:
Merge remote-tracking branch 'origin/main' into brushes

mattleibow committed:
Merge remote-tracking branch 'origin/main' into brushes

mattleibow committed:
Do the windows control color right

- don't use a "default", just clear
- map a few more label properties so they can be used to show brushes

mattleibow committed:
fix things

mattleibow commented on src/Core/src/Handlers/Label/LabelHandler.cs:
```diff
@@ -4,6 +4,9 @@ public partial class LabelHandler
 	{
 		public static PropertyMapper<ILabel, LabelHandler> LabelMapper = new PropertyMapper<ILabel, LabelHandler>(ViewHandler.ViewMapper)
 		{
+#if __IOS__ || IOS
```
I do know the old __IOS__ still exists, so we can do that. Or if we want to use the new IOS without underscores, we can also add that to the multitargets/compat projects

hartez commented:
Maui.Graphics has a similar cross-platform concept called Paint: https://github.com/dotnet/Microsoft.Maui.Graphics/blob/70431442b35b8ee6e46a0a7a471bc95ffadeeed9/src/Microsoft.Maui.Graphics/Paint.cs

Do we need to map Brushes -> Paint?

jsuarezruiz committed:
Merge branch 'main' into brushes

jsuarezruiz committed:
Merge branch 'main' into brushes

jsuarezruiz committed:
Fix build error

jsuarezruiz committed:
Implement Foreground Brush (replace TextColor)

jsuarezruiz committed:
Added Foreground support in WinUI

jsuarezruiz commented:
Added brushes support to other elements. Replaced `TextColor` property by `Foreground`:
<img width="1128" alt="maui-foreground" src="https://user-images.githubusercontent.com/6755973/115006370-14c44200-9ea9-11eb-8b97-a4e67cbc4a38.png">

jsuarezruiz committed:
Merge branch 'main' into brushes

jsuarezruiz committed:
Updated sample

jsuarezruiz committed:
Revert unnecessary changes

jsuarezruiz committed:
Fix build errors

jsuarezruiz committed:
Fix broken tests

jsuarezruiz committed:
Enable nullable in Windows BrushExtensions

jsuarezruiz committed:
Enable nullable in more classes

jsuarezruiz committed:
Merge branch 'main' into brushes

jsuarezruiz committed:
Fix build error

jsuarezruiz committed:
Updated sample

jsuarezruiz committed:
Update Foreground if text changes

jsuarezruiz committed:
Merge branch 'main' into brushes

jsuarezruiz commented:
Close this PR in favor of https://github.com/dotnet/maui/pull/892
